### PR TITLE
feat: disable setting COMPOSER_AUTH for gitlab

### PR DESCRIPTION
**Pull request by DjordyKoert**
## Changes
Adds new configuration option (inside `hostRules`) to disable setting COMPOSER_AUTH for gitlab.

New option: `artifactAuth`

## Context
https://github.com/renovatebot/renovate/discussions/20609

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/options/index.ts [hunk 1]
```diff
@@ -1998,6 +1998,7 @@ const options: RenovateOptions[] = [
     default: [],
     allowedValues: [
       'bundlerConservative',
+      'composerGitlabToken',
```
Comment (viceice) [inline lib/config/options/index.ts:2001]: @rarkins I'm not sure if this is the right way to solve it. 🤔


```suggestion
      'noComposerGitlabToken',
```
if we go this way, the value should reflect the behavior.

**Comment by DjordyKoert** [comment]
> I think this should be generic (no auth at all) and not gitlab specific

Wouldn't this make it possible to get rate-limited by github for example, when only the gitlab auth is not required?

**Comment by rarkins** [comment]
> > I think this should be generic (no auth at all) and not gitlab specific
> 
> Wouldn't this make it possible to get rate-limited by github for example, when only the gitlab auth is not required?

And wouldn't *that* give you `dist` entries for github.com, which you don't want? Or you are happy to have `dist` for `github.com` but not `dist` for `gitlab.com`?

**Comment by DjordyKoert** [comment]
> > > I think this should be generic (no auth at all) and not gitlab specific
> > 
> > 
> > Wouldn't this make it possible to get rate-limited by github for example, when only the gitlab auth is not required?
> 
> And wouldn't _that_ give you `dist` entries for github.com, which you don't want? Or you are happy to have `dist` for `github.com` but not `dist` for `gitlab.com`?

We are happy with receiving `dist` entries for github.com, we just do not need them for our own composer packages hosted on gitlab.

**Comment by rarkins** [comment]
I don't love this approach because next someone will pop up saying they're ok with gitlab.com but not with github.com, etc. Let's go back to the discussion to discuss more

File: lib/config/options/index.ts [hunk 2]
```diff
@@ -2244,6 +2244,19 @@ const options: RenovateOptions[] = [
     env: false,
     experimental: true,
   },
+  {
+    name: 'artifactAuth',
+    description:
+      'A list of package managers to enable artifact auth. Only managers on the list are enabled. All are enabled if `null`',
+    type: 'array',
```
Comment (rarkins) [inline lib/config/options/index.ts:2251]:
```suggestion
    experimental: true,
    type: 'array',
```

File: lib/modules/manager/composer/artifacts.ts [hunk 3]
```diff
@@ -55,7 +55,11 @@ function getAuthJson(): string | null {
   }
 
   hostRules.findAll({ hostType: 'gitlab' })?.forEach((gitlabHostRule) => {
-    if (gitlabHostRule?.token) {
+    if (
+      (!gitlabHostRule.artifactAuth ||
+        gitlabHostRule.artifactAuth.includes('composer')) &&
+      gitlabHostRule?.token
+    ) {
```
Comment (rarkins) [inline lib/modules/manager/composer/artifacts.ts:62]: This is too narrowly applying to just gitlab tokens, should apply to all hostRules used in this file

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/manager/composer/artifacts.ts (hunk 3 above)
Comment (DjordyKoert) [reply lib/modules/manager/composer/artifacts.ts]: Should this also be applied for github?

File: lib/modules/manager/composer/artifacts.ts [hunk 4]
```diff
@@ -55,6 +60,10 @@ function getAuthJson(): string | null {
   }
```
Comment (DjordyKoert) [inline lib/modules/manager/composer/artifacts.ts:60]: Should artifactAuth also be an available option for `github-oauth`?

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/composer/artifacts.ts [hunk 5]
```diff
@@ -27,34 +27,44 @@ import {
...
   );
+
   if (selectedGithubToken) {
     authJson['github-oauth'] = {
       'github.com': selectedGithubToken,
     };
   }
 
   hostRules.findAll({ hostType: 'gitlab' })?.forEach((gitlabHostRule) => {
+    if (!isArtifactAuthEnabled(gitlabHostRule)) {
```
[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/manager/composer/artifacts.ts:64]: we should convert to ordinary for-of loops for better code style

**Review by viceice — COMMENTED** [review]
see open discussions

**Comment by DjordyKoert** [comment]
> see open discussions

Missed this one, what discussion?

File: lib/modules/manager/composer/artifacts.ts [hunk 6]
```diff
@@ -27,34 +27,44 @@ import {
...
 } from './utils';
 
 function getAuthJson(): string | null {
   const authJson: AuthJson = {};
 
-  const githubToken = findGithubToken({
+  const githubHostRule = hostRules.find({
     hostType: 'github',
     url: 'https://api.github.com/',
   });
```
[the remark this answers is not in the document]
Comment (DjordyKoert) [reply lib/modules/manager/composer/artifacts.ts:41]: This was the easiest solution to implement that I could find without the need for a lot of refactoring.

If this is not the ideal solution, then recommendations would be appreciated 😄

**Review by viceice — APPROVED** [review]

**Review by secustor — APPROVED** [review]

**Comment by secustor** [comment]
@rarkins Could you take a look? Your review is blocking the merge.

**Review by rarkins — APPROVED** [review]

**Pull request merged**